### DISCUSSION TITLE
Can this be used with Playwright?

### DISCUSSION THREAD

**Discussion by waynerobinson** [FlareSolverr/FlareSolverr#856] [Q&A]
Is it possible to use FlareSolverr with Playwright?

I'm using the REST API get the cookies for a specific website and plugging them, along with the user agent into Playwright via the `context.addCookies()` method however, it looks like I'm still getting the Cloudflare Waiting Room for `https://digital.pnbank.com.au/` when trying to open that page. Both requests (FlareSolverr and Playwright) are coming from the same IP.

Is this even possible to use this way? If so, does someone have any idea what I'm doing wrong?

Could this project be changed to include the developer tools URL or similar of a session? I *think* that could be used with Playwright or Puppeteer to remote control the browser that just solved the captcha?

If we could get the CDP URL from webdriver we could then connect in Playwright via:

```js
const browser = await playwright.chromium.connectOverCDP('http://path-to-the-devtools-server:PORT');
const defaultContext = browser.contexts()[0];
const page = defaultContext.pages()[0];
```

**Comment by ngosang** [comment]
You need to change the Playwright user-agent to match the Flaresolverr user-agent. If its not possible you can use Python.

**Reply by waynerobinson** [reply]
As mentioned, I used the same user agent, but to no avail. I suspect that they’re doing TLS fingerprinting or similar.

Which is why it would be nice to just be able to connect to the Chrom(e/ium) session that’s created by FlareSolverr.

**Reply by ngosang** [reply]
They do TLS fingerprinting during the challenge, but when you have the cookie they only chech the user-agent. You should open a CheckMyUserAgent with Playwrigt to make sure you are changing the user agent.

An alternative could be using just FlareSolverr to make all requests. Check about sessions in the readme.

**Reply by waynerobinson** [reply]
Thanks for that! Seems that `puppeteer-extra-plugin-stealth` was overriding the set user agent.

**Comment by yaracil** [comment]
Were you able to find out the 'path-to-the-devtools-server' to do the cdp connection?

**Comment by peipeiwei303-netizen** [comment]
Yes, this is technically possible, but the issue you’re hitting is not Playwright vs FlareSolverr integration — it’s session consistency across the full request chain.

What’s happening in your flow is usually:

FlareSolverr solves the challenge → issues cookies
Playwright reuses cookies → but request still looks like a “new client context”
Same IP + slightly different browser fingerprint → Cloudflare re-evaluates risk → Waiting Room again

So even though the cookies are valid, the *combined identity* is not stable.

A more reliable pattern is:

1. Keep a single consistent browser identity (Playwright context or CDP session)
2. Solve challenge inside that same session (not external solver → import cookies)
3. Ensure IP stability during the entire session lifecycle

If FlareSolverr is used, it should ideally:

   operate on the same proxy exit IP
   or return a session that can be directly “continued” via CDP rather than replayed via cookies

Playwright CDP (`connectOverCDP`) can work well here, but only if:

   the session origin (IP + TLS + browser fingerprint) remains identical
   you avoid switching between solver runtime and automation runtime mid-session

In practice, people often combine this with residential proxy infrastructure (e.g. Novada-style sticky sessions) so that:

   CAPTCHA solving and browsing happen on the same IP
   session continuity is preserved across navigation
   secondary challenges are reduced because the identity doesn’t “jump”

Without that, Cloudflare tends to treat the cookie reuse as a partial replay attack rather than a continuous human session.
